zilliqa/src/consensus.rs

```diff
@@ -1128,59 +1380,69 @@ impl Consensus {
         };
         // Prevent the creation of forks from the already committed chain
         if block_high_qc_block.view() < finalized_block.view() {
-            return Err(anyhow!("invalid block"));
+            return Err(anyhow!(
+                "invalid block - high QC view is {} while finalized is {}",
+                block_high_qc_block.view(),
+                finalized_block.view()
+            ));
         }
 
         // This block's timestamp must be greater than or equal to the parent block's timestamp.
         if block.timestamp() < parent.timestamp() {
             return Err(anyhow!("timestamp decreased from parent"));
         }
 
-        // This block's timestamp must be at most `self.allowed_timestamp_skew` away from the current time. Note this
+        // This block's timestamp should be at most `self.allowed_timestamp_skew` away from the current time. Note this
         // can be either forwards or backwards in time.
+        // Genesis is an exception for now since the timestamp can differ across nodes
         let difference = block
             .timestamp()
             .elapsed()
             .unwrap_or_else(|err| err.duration());
-        if difference > self.config.allowed_timestamp_skew {
+        if difference > self.config.allowed_timestamp_skew && parent.view() > 0 {
```

Comment:
IMO, the timestamp logic should just be 'is it more than the previous timestamp' - and maybe if it is a block proposal, is it less than now.

The problem you were seeing @JamesHinshelwood  about timestamp skew is that if you sync the chain, then blocks you check to be 'safe' fail this check since they were minted ages ago.

Reply:
I think timestamp skew should prevent blocks being mined too far into the future (to allow for e.g. ~1-2s skew, at most, between system clocks in nodes), but I agree that it shouldn't restrict blocks in the past as long as they're later than their parent block.